kart/add_dataset.py

```diff
@@ -69,6 +73,14 @@ def add_dataset(ctx, table_name, message, launch_editor, output_format):
         table_name, meta_items
     )
 
+    # Create a table object from the table_name:
+    path = TableV3.dataset_path_to_table_name(table_name)
+    table = TableV3.new_dataset_for_writing(path, meta_items["schema.json"], repo)
+
+    # Add triggers on a new table if they aren't already present:
+    with wc.session() as sess:
```

Comment:
there's a race condition we should think about:

* user runs `kart add-dataset` on a table with 10 rows
* we build a diff
* user inserts a new row
* then we create triggers
* then we commit 10 rows to the dataset

This sequence of events causes the triggers to miss an insert - the row is not tracked by the tracking table, nor is it committed. User won't see a diff until they change the row and it gets picked up by the trigger again (it will be considered an update rather than an insert, even though the row hasn't been committed)

What if instead, we created the triggers before we fetch the diff?

* user runs `kart add-dataset` on a table with 10 rows
* we create triggers
* user inserts a new row
* trigger tracks a new inserted row
* we fetch a diff of the whole table
* we commit the diff

In this situation^ we have a different issue - the row is _both_ committed and in the tracking table - in both cases as an 'insert'. I'm not sure if this causes problems or not. I'd hope it doesn't show up in `kart diff` but I might be mistaken.


@olsen232 is this something we should address?

Reply:
Having unneeded primary-key-values in the tracking table isn't that bad - they don't cause the diffs to contain any extra deltas, they just slightly add to Kart's workload (in the extreme case, we could put every row in the tracking table for no reason, kart would still work fine, but it would lose the efficiency that the tracking table provides). So, creating triggers early is preferable to creating them late.

Apart from that preference, I wouldn't put too much more thought into it... all of this probably happens inside a single transaction, so, I don't know if reordering the pieces of it will make any difference, or if it all appears atomic to another editor. We could test it further but it seems like a lot of work for a weird edge case